Sources/SwiftFrameCore/Workers/ImageWriter.swift

```diff
@@ -0,0 +1,103 @@
+import AppKit
+import Foundation
+
+public final class ImageWriter {
+
+    // MARK: - Exporting
+
+    static func finish(
+        context: CGContext,
+        with outputPaths: [FileURL],
+        sliceSize: CGSize,
+        outputWholeImage: Bool,
+        locale: String,
+        suffix: String,
+        format: FileFormat,
+        completion: @escaping () throws -> Void) throws {
+        guard let image = context.makeImage() else {
+            throw NSError(description: "Could not render output image")
+        }
+        DispatchQueue.global().ky_asyncOrExit {
+            let slices = sliceImage(image, with: sliceSize)
+            var slicesFinished = false
+            var bigImageFinished = !outputWholeImage
+
+            let workGroup = DispatchGroup()
+            workGroup.enter()
+
+            // Writing images asynchronously gave a big performance boost, what a surprise
+            // Also, since we checked beforehand if the directory is writable, we can safely put of the rendering work to a different queue
+            DispatchQueue.global().ky_asyncOrExit {
+                try ImageWriter.write(images: slices, to: outputPaths, locale: locale, suffix: suffix, format: format)
+                slicesFinished = true
+                if slicesFinished && bigImageFinished {
+                    workGroup.leave()
+                }
+            }
+
+            if outputWholeImage {
+                DispatchQueue.global().ky_asyncOrExit {
+                    try outputPaths.forEach {
+                        try ImageWriter.write(image, to: $0.absoluteURL.appendingPathComponent(locale), fileName: "\(locale)-\(suffix)-big", format: format)
+                    }
+                    bigImageFinished = true
+                    if slicesFinished && bigImageFinished {
+                        workGroup.leave()
+                    }
+                }
+            }
+
+            _ = workGroup.wait(timeout: .now() + 5.00)
```

Comment:
I mentioned this before. I think making assumptions about timing is tricky. Depending on system load and other potentially ongoing file system operations, maybe it could take longer than 5 seconds. I think it's ok to not set a time out. If the process is really stuck, you can always kill it from the CLI.

PS: I think `.now() + .seconds(5)` would be the slightly safer approach if you wanted to have a timeout because then you don't have to worry about units.